lms/djangoapps/instructor/views/api.py

```diff
@@ -345,27 +351,100 @@ def register_and_enroll_students(request, course_id):  # pylint: disable=too-man
 
-            # Iterate each student in the uploaded csv file.
-            email = student[EMAIL_INDEX]
-            username = student[USERNAME_INDEX]
-            name = student[NAME_INDEX]
-            country = student[COUNTRY_INDEX][:2]
+            if extra_fields_is_enabled:
+                is_valid_csv = 4 <= len(student) <= 6
+                error = _('Data in row #{row_num} must have between four and six columns: '
+                          'email, username, full name, country, cohort, and course mode. '
+                          'The last two columns are optional.').format(row_num=row_num)
+            else:
+                is_valid_csv = len(student) == 4
+                error = _('Data in row #{row_num} must have exactly four columns: '
+                          'email, username, full name, and country').format(row_num=row_num)
+
+            if not is_valid_csv:
+                general_errors.append({
+                    'username': '',
+                    'email': '',
+                    'response': error
+                })
+                continue
+
+            # Extract each column, handle optional columns if they exist.
+
+            email, username, name, country, *optional_cols = student
+            if optional_cols:
+                optional_cols.append(default_course_mode)
+                cohort_name, course_mode, *_tail = optional_cols
+            else:
+                cohort_name = None
+                course_mode = None
+
+            # Validate cohort name, and get the cohort object.  Skip if course
+            # is not cohorted.
+
+            cohort = None
+
+            if cohort_name and not already_warned_not_cohorted:
+                if not is_course_cohorted(course_id):
+                    row_errors.append({
+                        'username': username,
+                        'email': email,
+                        'response': _('Course is not cohorted but cohort provided. '
+                                      'Ignoring cohort assignment for all users.')
+                    })
+                    already_warned_not_cohorted = True
+                elif cohort_name in cohorts_cache:
+                    cohort = cohorts_cache[cohort_name]
+                else:
+                    cohort = CourseUserGroup.objects.filter(
+                        course_id=course_id,
+                        group_type=CourseUserGroup.COHORT,
+                        name=cohort_name
+                    ).first()
+                    # Don't attempt to create cohort or assign student if cohort
+                    # does not exist.
+                    if cohort is None:
+                        row_errors.append({
+                            'username': username,
+                            'email': email,
+                            'response': _('Cohort name not found: {cohort}. '
+                                          'Ignoring cohort assignment.').format(cohort=cohort_name)
```

Comment:
This is going to appear only once per cohort - in the subsequent runs the cohort will already be cached, so it'll execute the `elif`. Let's make this obvious that we're ignoring all cohort assignments for this specific cohort.

If we would like to display this message for each assignment, then we should merge `elif` and `else`.

Reply:
@Agrendalath:

> Let's make this obvious that we're ignoring all cohort assignments for this specific cohort.

Can you be more specific, in particular, what exactly you think would be an obvious way to communicate we are ignoring all cohort assignments for this specific cohort and to who (e.g. user, developer, both)?

Reply:
@jvdm, sure - this is the message for users who import the CSV. When they're getting this error message, they see the following data: `username`, `email`, `error message`. Therefore, it should be clear for them that this error doesn't only apply to the specified user, but to the whole cohort with the `{cohort}` name. This is a similar case to the previous error message, which is displayed when the course is not cohorted. Its last sentence is "Ignoring cohort assignment for all users.". Therefore, we can rephrase this message to something along "Ignoring all assignments for this cohort.".

Reasoning:
The `elif cohort_name in cohorts_cache` check is going to be `False` only once per cohort name because the next loops are going to have this cohort cached.